Add a modified infiltration scheme in ELM to improve surface water dynamics

donghuix opened the pull request:
In default, the infiltration rate is the same for the non-water surface and surface water storage within the grid cell. This assumption results in a significant underestimation of surface water because the infiltration in the surface water storage is not constrained. 

In this PR, we introduced a modified infiltration scheme to constrain the infiltration in the surface water storage. The infiltration rate in the surface water storage is smaller than the infiltration rate in the non-water surface. More details about this new scheme and corresponding evaluations can be found in the preprint: https://doi.org/10.21203/rs.3.rs-2733749/v1 The manuscript is still under review. 

The new infiltration scheme can be turned by adding ``use_modified_infil = .true.`` in ``user_nl_elm``. 

We passed all the tests in `e3sm_land_developer` test suite.

We also created a test for the simulation of surface water dynamics with the new infiltration scheme and calibrated parameter for NLDAS domain: ``surface_water_dynamics``. This test was added into ``e3sm_land_developer`` test suite. 

[BFB]

bishtgautam commented on components/elm/src/biogeophys/SoilHydrologyType.F90:
```diff
@@ -542,18 +547,36 @@ subroutine InitCold(this, bounds)
+    call ncd_pio_openfile (ncid, locfn, 0)
+    call ncd_io(ncid=ncid, varname='pc', flag='read', data=this%pc, dim1name=grlnd, readvar=readvar)
+    if (.not. readvar) then
+      this%pc(:) = 0.4
```
Change `0.4` to `0.4_r8`

donghuix replied:
Using `0.4_r8` will result in non BFB to the master branch. That is because `pc` is set to 0.4 in the master (Please see next reply).

bishtgautam replied:
In that case, leave it as is. Thanks.

bishtgautam commented on components/elm/src/external_models/emi/src/elm_stub/utils/clm_varcon.F90:
```diff
@@ -43,7 +43,7 @@ module clm_varcon
   real(r8), parameter :: n_melt=0.7     !fsca shape parameter
   real(r8), parameter :: e_ice=6.0      !soil ice impedance factor
-  real(r8), parameter :: pc = 0.4       !threshold probability
+  !real(r8), parameter :: pc = 0.4       !threshold probability
```
please delete this line

bishtgautam commented on components/elm/src/main/elm_varcon.F90:
```diff
@@ -42,7 +42,7 @@ module elm_varcon
   real(r8), parameter :: n_melt=0.7                         ! fsca shape parameter
   real(r8), parameter :: e_ice=6.0                          ! soil ice impedance factor
-  real(r8), parameter :: pc = 0.4                           ! threshold probability
+  !real(r8), parameter :: pc = 0.4                           ! threshold probability
```
please delete it.

donghuix replied:
Ok.

bishtgautam commented on components/elm/src/biogeophys/SoilHydrologyMod.F90:
```diff
@@ -514,7 +549,17 @@ subroutine Infiltration(bounds, num_hydrologyc, filter_hydrologyc, num_urbanc, f
+                if ( use_modified_infil ) then
+                   ! Assume frac_h2osfc occurs on top of fsat
+                   if (frac_h2osfc(c) <= fsat(c)) then
+                     qflx_h2osfc_drain(c)=0
```
Change `0` to `0._r8`.

donghuix replied:
Ok.

bishtgautam reviewed (approved):
@donghuix I have a few minor comments for you to address.

peterdschwartz commented:
Are all the input files for the test on the input data server? Will `use_modified_infil = .true.` work without needing calibrated parameters?  If not, I would suggest making a compset or use case to take care of all the dependencies.

Also, urls can't go into the git log, so do you have a citation that could be used instead?

donghuix commented:
@bishtgautam I addressed your comments in the new commit. However, I cannot add `._r8` for the assignment of `pc`, which will result in differences compared to baseline simulation.

@peterdschwartz Yes, @bishtgautam has uploaded a calibrated ELM surface dataset to the input data server for the test case. @bishtgautam Can you confirm it?

`use_modified_infil = .true.` doesn't need calibrated parameter, and it works with any ELM surface dataset as default parameter values will be used. 

I changed the urls to citation in the log. Please let me know if additional modifications are needed. 

Thanks!

peterdschwartz commented:
> I changed the urls to citation in the log. Please let me know if additional modifications are needed.
> 
> Thanks!

@donghuix I should have been more precise.  When I merge your PR, I'll copy/paste your PR description at the top of this page as the merge commit message. So, I just need this researchgate url replaced with the citation
> In this PR, we introduced a modified infiltration scheme to constrain the infiltration in the surface water storage. The infiltration rate in the surface water storage is smaller than the infiltration rate in the non-water surface. More details about this new scheme and corresponding evaluations can be found in the preprint: https://www.researchsquare.com/article/rs-2733749/v1. The manuscript is still under review.

Thanks

rljacob commented:
You can use the DOI  https://doi.org/10.21203/rs.3.rs-2733749/v1

peterdschwartz commented:
> @peterdschwartz Yes, @bishtgautam has uploaded a calibrated ELM surface dataset to the input data server for the test case. @bishtgautam Can you confirm it?

Tried to run the new test on chrysalis and the input file is not available for download.....  @donghuix @bishtgautam

peterdschwartz commented:
Tests show DIFFs in these tests:
```
 SMS_Ln9.ne4_oQU240.F2010.chrysalis_intel.eam-outfrq9s (Overall: DIFF) details:
  SMS_Ly2_P1x1.1x1_smallvilleIA.IELMCNCROP.chrysalis_intel.elm-force_netcdf_pio (Overall: DIFF) details:
  SMS_Ly2_P1x1.1x1_smallvilleIA.IELMCNCROP.chrysalis_intel.elm-per_crop (Overall: DIFF) details:
  SMS_Ly2_P1x1_D.1x1_smallvilleIA.IELMCNCROP.chrysalis_intel.elm-lulcc_sville (Overall: DIFF) details
```
Appear to be only due to the new history field added. Could put an if statement around the new history field addition with the new namelist flag ?

donghuix commented:
> Tests show DIFFs in these tests:
> 
> ```
>  SMS_Ln9.ne4_oQU240.F2010.chrysalis_intel.eam-outfrq9s (Overall: DIFF) details:
>   SMS_Ly2_P1x1.1x1_smallvilleIA.IELMCNCROP.chrysalis_intel.elm-force_netcdf_pio (Overall: DIFF) details:
>   SMS_Ly2_P1x1.1x1_smallvilleIA.IELMCNCROP.chrysalis_intel.elm-per_crop (Overall: DIFF) details:
>   SMS_Ly2_P1x1_D.1x1_smallvilleIA.IELMCNCROP.chrysalis_intel.elm-lulcc_sville (Overall: DIFF) details
> ```
> 
> Appear to be only due to the new history field added. Could put an if statement around the new history field addition with the new namelist flag ?

I removed the new variable in the history field, as it is not needed. Let me know if it passes the test now.

peterdschwartz commented:
Yes it looks like the DIFFs are resolved.

peterdschwartz commented:
merged to master.  Will wait till tomorrow to bless pm test
